pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -64,22 +67,30 @@ pub async fn download_layer_file<'a>(
             // TODO: this doesn't use the cached fd for some reason?
             let mut destination_file = fs::File::create(&temp_file_path).await.with_context(|| {
                 format!(
-                    "Failed to create a destination file for layer '{}'",
+                    "create a destination file for layer '{}'",
                     temp_file_path.display()
                 )
             })
             .map_err(DownloadError::Other)?;
             let mut download = storage.download(&remote_path).await.with_context(|| {
                 format!(
-                    "Failed to open a download stream for layer with remote storage path '{remote_path:?}'"
+                    "open a download stream for layer with remote storage path '{remote_path:?}'"
                 )
             })
             .map_err(DownloadError::Other)?;
-            let bytes_amount = tokio::io::copy(&mut download.download_stream, &mut destination_file).await.with_context(|| {
-                format!("Failed to download layer with remote storage path '{remote_path:?}' into file {temp_file_path:?}")
-            })
-            .map_err(DownloadError::Other)?;
+
+            let fut = tokio::io::copy(&mut download.download_stream, &mut destination_file);
+
+            let bytes_amount = timeout(Duration::from_secs(120), fut)
```

Comment:
We shouldn't use timeout on the copy operation as a whole, but rather, on the individual (many) read and write calls that the operation does under the hood.

Copy from my comment in https://github.com/neondatabase/neon/issues/3672#issuecomment-1439956748:

> 
> Regardless of pro & con: if we decide to add a timeout, it must not be on the copy operation as a whole, but the read & write operations invoked by copy.
> I.e., not like this:
> 
> tokio::time::timeout(Duration::from_secs(30), tokio::io::copy(...))
> 
> because, if the download is a little slow, or the disk is a little slow, but is making progress, we'd still give up after 30 seconds, waste all that work, and likely retry the download shortly after, because something clearly needs this layer file.
> 
> IDK if there is a NoActivityPipe type that we could interpose here.

Reply:
> but rather, on the individual (many) read and write calls that the operation does under the hood
Why ? we are putting here a safety net of two minutes which is large enough to withstand time pressure operations
It can even be set to more, but the idea here is "un" stuck the system.
While a complete system hang allows you for easier debugging having the timeouts will allow the system to resurrect and continue working

Reply:
>  but the idea here is "un" stuck the system.

We're on the same page here.

What I'm saying is that, to decide whether the system is stuck, one shouldn't use a plain timeout, which is ultimately system-load- and hardware-dependent, but instead, have a timeout that resets whenever some progress was made.
In this case, "progress was made <=> (bytes were read from download stream or bytes were written to destination file)".

Reply:
I don't follow
even if we had
while (!eof) {
copy(src,dst) for small chuncks
}
we still want to time out

Reply:
Would the wrappers provided by tokio-io-timeout be what @problame is after? If I understand correctly, those reset on progress: https://docs.rs/tokio-io-timeout/latest/tokio_io_timeout/struct.TimeoutWriter.html

I thought we already had this dependency, but I guess not.

Reply:
just to make sure I follow
wrap the Download.download_stream with a TimeoutReader 
and take from the destination file its stream and wrap it in TimeoutWriter ?

Reply:
I'm not sure its worth the complication.

>  if the download is a little slow, or the disk is a little slow

If download takes more than two minutes IMO something is really off, not just a little slow. Actually breaking the loop and starting over may be faster than waiting for every couple of bytes up to two minutes. 

Reminds me about Happy Eyeballs algorithm https://en.wikipedia.org/wiki/Happy_Eyeballs, may be a good idea to implement something like that. I e if download is slow, start new one and pick whichever completes first. Though I understand that this will use more resources, can possibly drive us to exceeding aws limits, etc

Reply:
`TimeoutReader` and `TimeoutWriter` implements exactly what I was proposing.

> just to make sure I follow
> wrap the Download.download_stream with a TimeoutReader
> and take from the destination file its stream and wrap it in TimeoutWriter ?

@shanyp  yes, precisely that.

---

To @LizardWizzard 's point: I know that problem from downloading Linux ISOs where sometimes you hit an overloaded mirror, and by retrying, the load balancing mechanism will likely give you another one.
But, is that really an issue with AWS in practice?
I'd rather assume the problem is on our end than on theirs, in the majority of cases.
Could be different for different deployment / cloud provider.

Regarding

> I'm not sure its worth the complication.

Well, the effort is really low, I think. Drop in the dependency, use it, done.
Or what type of 'complication' are you referring to here?

Reply:
> But, is that really an issue with AWS in practice?

This is what they recommend for latency sensitive applications. See https://docs.aws.amazon.com/pdfs/whitepapers/latest/s3-optimizing-performance-best-practices/s3-optimizing-performance-best-practices.pdf Paragraph `Retry Requests for Latency-Sensitive Applications`

> Well, the effort is really low, I think

If it is indeed what we heed, then yes. But from my point it also increases the mental load when one is reading this code. Timeout for whole operation is the simplest option.

> Drop in the dependency

I'm not opposed to adding new dependencies in general, but if default timeout combinator works for us then IMO its not worth it. One more plus is that we're not adding to compilation times. Though the dependency is really small, so it shouldnt really make a difference in practice. On the other side if we start to add many such libraries from time to time the effect can become noticeable